src/rabbit_access_control.erl

```diff
@@ -178,17 +178,18 @@ check_access(Fun, Module, ErrStr, ErrArgs) ->
     check_access(Fun, Module, ErrStr, ErrArgs, access_refused).
 
 check_access(Fun, Module, ErrStr, ErrArgs, ErrName) ->
-    Allow = case Fun() of
-                {error, E}  ->
-                    rabbit_log:error(ErrStr ++ " by ~s: ~p~n",
-                                     ErrArgs ++ [Module, E]),
-                    false;
-                Else ->
-                    Else
-            end,
-    case Allow of
+    case Fun() of
         true ->
             ok;
         false ->
-            rabbit_misc:protocol_error(ErrName, ErrStr, ErrArgs)
+            rabbit_misc:protocol_error(ErrName, ErrStr, ErrArgs);
+        {error, E}  ->
+            rabbit_log:error(ErrStr ++ " by ~s: ~p~n",
+                             ErrArgs ++ [Module, E]),
+            rabbit_misc:protocol_error(ErrName, ErrStr, ErrArgs);
+        {error_message, Message} ->
```

Comment:
`{error, Message}` can return an arbitrary error reason and will only log it. `{error_message, Message}` will also add it to the protocol error sent to a client.